### PR TITLE
search contexts: support searching by spec

### LINKED ISSUE #21042: search contexts: support searching by spec
Searching for contexts in dropdown and list page should support searches by spec. E.g. `@rok/release`, should search for contexts with namespace name containing `rok` and context names containing `release`.

### PR DESCRIPTION
**Pull request by novoselrok**
Fixes #21042

### CODE REVIEW

**Comment by novoselrok** [comment]
@camdencheek -> https://github.com/sourcegraph/sourcegraph/pull/21290/commits/1a70f5d7d7724a7edd7ea5598d46267aa1b7a206

@limitedmage -> https://github.com/sourcegraph/sourcegraph/pull/21290/commits/12dd7bcf4b14a6e6c3255e055f0238b6f46eff2b

**Review by camdencheek — APPROVED** [review]

File: client/web/src/search/input/HighlightedSearchContextSpec.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    if (index === -1) {
+        return <>{text}</>
+    }
+    const before = text.slice(0, index)
+    const highlighted = text.slice(index, index + highlightPart.length)
+    const after = text.slice(index + highlightPart.length)
+    return (
+        <>
+            {before}
+            <strong className={highlightClass}>{highlighted}</strong>
```
Comment (limitedmage) [inline client/web/src/search/input/HighlightedSearchContextSpec.tsx:32]: **very minor:** semantically, I think `mark` might make more sense than `strong` (we could still make it look bold visually with CSS)

File: client/web/src/search/input/HighlightedSearchContextSpec.tsx [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+export const HighlightedSearchContextSpec: React.FunctionComponent<{ spec: string; searchFilter: string }> = ({
+    spec,
+    searchFilter,
+}) => {
+    if (searchFilter.length === 0) {
+        return <>{spec}</>
+    }
+    // Parse search filter as a search context spec and
+    // highlight both parts (namespace and name) separately
```
Comment (limitedmage) [inline client/web/src/search/input/HighlightedSearchContextSpec.tsx:54]: What happens if I type part of a fully-qualified name in the search, eg. `user/test` without the `@`?

**Review by limitedmage — APPROVED** [review]

File: client/web/src/search/input/HighlightedSearchContextSpec.tsx (hunk 2 above)
Comment (novoselrok) [reply client/web/src/search/input/HighlightedSearchContextSpec.tsx:54]: Then it uses the query to search just through context names. `user/test` is also a valid context name so we can't assume it's namespaced.

I realise it's not ideal, but I really wouldn't want to go down a path where we handle a ton of arbitrary input.

//cc @quinnkeast

File: client/web/src/search/input/HighlightedSearchContextSpec.tsx (hunk 1 above)
Comment (novoselrok) [reply client/web/src/search/input/HighlightedSearchContextSpec.tsx:32]: `mark` also adds a yellowish background to the highlighted parts. @quinnkeast would that be ok?

Screenshot 2021-05-25 at 08 02 26

Comment (quinnkeast) [reply client/web/src/search/input/HighlightedSearchContextSpec.tsx:32]: Preferably not highlighted (for now, maybe we'll change it systematically in the redesign).

**Review by quinnkeast — APPROVED** [review]

**Pull request merged**